### ISSUE TITLE
Question about flows: how would you bind the width of an element to an external slider ?

### ISSUE THREAD

**Issue by nferruzzi** [apptekstudios/ASCollectionView#42]
Hi,
thanks for the package! I'm evaluating it and at the moment work like a charm (on both iOS and catalyst). I've been trying to make a custom layout pinned to an external @State .. something like this based on your `PhotoGridScreen`

```
    @State private var minWidth: CGFloat = 0.5

...
     Slider(value: self.$minWidth, in: 0.1...1.0, step: 0.01)
...

    var widthLayout: ASCollectionLayout<Int> {
        debugPrint("Called with \(self.minWidth)")
        return ASCollectionLayout(scrollDirection: .vertical, interSectionSpacing: 0) {
            ASCollectionLayoutSection { environment in
                let width = environment.container.effectiveContentSize.width

                let itemWidth = environment.container.effectiveContentSize.width * self.minWidth

                /// This closure is called when the minWidth changes but it always see the "initial" value
                /// it changes only when I do resize. Funny enough it jumps back to 0.5 when I move the slider after resizing the parent view
                 debugPrint("Min width: \(self.minWidth) Item width: \(itemWidth)")

                let gridItemInsets = NSDirectionalEdgeInsets(top: 5, leading: 5, bottom: 5, trailing: 5)
                let itemSize = NSCollectionLayoutSize(widthDimension: .absolute(itemWidth),
                                                      heightDimension: .absolute(itemWidth))

                let item = NSCollectionLayoutItem(layoutSize: itemSize)
                item.contentInsets = gridItemInsets

                let count = Int(width / itemWidth)

                let rowGroupSize = NSCollectionLayoutSize(widthDimension: .absolute(itemWidth * CGFloat(count)),
                                                          heightDimension: .absolute(itemWidth))
                let rowGroup = NSCollectionLayoutGroup.horizontal(layoutSize: rowGroupSize, subitem: item, count: count)

                let section = NSCollectionLayoutSection(group: rowGroup)
                return section
            }
        }
    }
```

but the layout is not refreshed when I move the slider, I also need to resize the parent view (I've built the Demo for catalyst). So the question is, how can I achieve something like this?

an output example 
```
"Called with 0.5"
"Binding<CGFloat>(transaction: SwiftUI.Transaction(plist: []), location: SwiftUI.StoredLocation<CoreGraphics.CGFloat>, _value: 0.5) Min width: 0.5 Item width: 395.75"
"Binding<CGFloat>(transaction: SwiftUI.Transaction(plist: []), location: SwiftUI.StoredLocation<CoreGraphics.CGFloat>, _value: 0.5) Min width: 0.5 Item width: 395.75"
"Called with 0.53"
"Binding<CGFloat>(transaction: SwiftUI.Transaction(plist: []), location: SwiftUI.StoredLocation<CoreGraphics.CGFloat>, _value: 0.5) Min width: 0.5 Item width: 395.75"
```
thanks in advance

**Comment by nferruzzi** [comment]
Can it be that the closure result is cached somehow? because sometimes I can see the value being the initial one (minWidth: 0.5) even if the slider has a different value

**Comment by apptekstudios** [comment]
Thanks for trying out the library @nferruzzi . This behaviour is actually expected by default, although this certainly is a use case I’d like to support. At current it seems UICollectionViewCompositionalLayout doesn’t allow changing the sectionProvider after it is initialised. This means we would need to recreate the layout object every time the view is updated to keep the closure in sync. However, it should be relatively straightforward to add a way to request that ASCollectionView recreate the layout when a certain value is updated. It’s late here in Australia but hopefully I get time to experiment tomorrow :)

**Comment by apptekstudios** [comment]
Quickly implemented a demo of how to achieve this. It did require adding a setting to ASCollectionView, and I'd like to give the implementation / API a little more thought before moving into master, so just a branch for now:
https://github.com/apptekstudios/ASCollectionView/tree/layoutInvalidation

If you download this branch and run the demo project (go to 'Experimental: adjustable layout) you can see it in action 😄 

Note:
- Storing the layout relevant state in a class, such that you can use it in the layout closure and have it stay in sync 👍

**Comment by nferruzzi** [comment]
thank you! the demo works great, I've also noticed that it doesn't make a difference if `shouldInvalidateLayoutOnStateChange` is true or false, from my tests it always recreates the proper layout; is this an expected behavior or a misunderstanding on my side of the property?

BUT: no matter what I do, when I recreate the same structure on my project (state class, layout, stepper) it doesn't work anymore XD. The only different thing is the model, but everything else is the same.

I'm getting crazy over it >.<
Anyway thank you for you demo, I'll be waiting for the merge  :)

**Comment by apptekstudios** [comment]
Is your class an ObservableObject and referenced as `@ObservedObject`? If a class is referenced as `@State` it won't trigger a reload of the view.

Otherwise the only thing I can think is to check: does it respond as expected if you rotate the device? (because that forces the layout to recalculate)

**Comment by apptekstudios** [comment]
Regarding the layout updating regardless of `shouldInvalidateLayoutOnStateChange`:
- that property simply ensures that the layout does get invalidated
- quite often the layout object will actually invalidate itself automatically when the data is updated.

**Comment by tarasis** [comment]
I'm grateful you are working on this, and produced this experiment branch so quickly. This is exactly the feature that was missing for me from this library.

I had been using a different gridding library which was fine with a handful of cells, but after a few hundred became useless. I tried a bunch of others yesterday with largely similar results.

Today I just tried this one and it worked great. No more 6 second delay switching back to the tab :)

**Comment by tarasis** [comment]
> BUT: no matter what I do, when I recreate the same structure on my project (state class, layout, stepper) it doesn't work anymore XD. The only different thing is the model, but everything else is the same.

By chance is it not compiling? I think I'm hitting the same thing. My data source (array of a class) works fine when supplied to ASCollectionView, but when I try and use It with ASCollectionViewSection it reports an error: 

"Cannot convert value of type '[MyClassName]' to expected argument type '[_]'"

**Comment by nferruzzi** [comment]
No mine compiles fine is just it doesn't update in realtime like the demo project. When I replace my View containing the ASCollectionView with the one of this experimental branch it doesn't work either so its not the layout for the model is something not correlated at all.  I think I'm hitting some weirdness of SwiftUI because I found out after hours of tests that my problem is linked to the presence on a parent view of an @EnvironmentObject that for unknow reason block that sliding behavior. I've tried to reproduce the same hierarchy in the Demo project but without any luck. I've to say that the past two weeks working with apple techs have been a nightmare, 3 crashes out of our control (dead_code, uitextview and now the problem with 10.15.1). This one just looks like an "ordinary" day of not working stuff XD

**Comment by nferruzzi** [comment]
@tarasis about your problem, I've seen many times the compiler pointing out errors in the wrong place. Especially with the complex types inference of swiftui. The problem you have maybe a typo somewhere else, double check your code commenting out sections, this is what I have to do way too often.

**Comment by tarasis** [comment]
> @tarasis about your problem, I've seen many times the compiler pointing out errors in the wrong place. Especially with the complex types inference of swiftui. The problem you have maybe a typo somewhere else, double check your code commenting out sections, this is what I have to do way too often.

Thanks, yeah, I was passing a string to ASRemoteImageView rather than a URL, but the compiler was perfectly happy with the line, and was saying it was an issue with the data model.

**Comment by tarasis** [comment]
Now working, thanks @nferruzzi. 

Is it possible to have the layout reflect the change to the number of columns as the datasource is being loaded?

From what I see the change to the number of columns by the stepper (or slider) won't happen until the datasource has finished loading & I swap away from the view and back to it (switch to a different tab and then back).

The same is true if I swap to the tab, wait for the data to finish loading with the ASCollectionView fully populated, and then try to change the number of columns. I have to change away from the tab once, and then back again and now changes to the stepper affect the ASCollectionView.

If I wait for the data source to fully load, then go to the tab containing the ASCollectionView and then change the stepper, the number of columns changes correctly.

**Comment by apptekstudios** [comment]
@tarasis Are you finding the above issue in the demo project or with another project? It would be fantastic if you're able to reproduce it in a demo so I can troubleshoot :)
It sounds to me as though your layout closure may be capturing a value type (as opposed to a class type). Add a print statement to your layout closure and check whether A) it is being called when you change the stepper, and B) the value that is then used in the closure is correct (and not the initial value from view creation)

@nferruzzi Any luck figuring out what's going on with your project?
- Try putting a breakpoint in the `updateLayout()` function of ASCollectionView.Coordinator and check that it is actually being called when your stepper/slider changes.
- Are you using compositional layout or magazine layout at the moment? It occurred to me that this might be a difference between the demo and your project

**Comment by tarasis** [comment]
> @tarasis Are you finding the above issue in the demo project or with another project? It would be fantastic if you're able to reproduce it in a demo so I can troubleshoot :)
> It sounds to me as though your layout closure may be capturing a value type (as opposed to a class type). Add a print statement to your layout closure and check whether A) it is being called when you change the stepper, and B) the value that is then used in the closure is correct (and not the initial value from view creation)

@apptekstudios this issue is in my project. Will try and sort a demo out for you. (For the other issue I just separately raised, I've seen in my app but have also recreated in your demo app with a hack to add a second tab)

**Comment by apptekstudios** [comment]
@nferruzzi 
I realised perhaps SwiftUI is not calling the `updateUIViewController` method as it doesn't think the LayoutInfo class is relevant to ASCollectionView. Can you put a breakpoint in `updateUIViewController` and see if it is called when you change the stepper?

If so, try wrapping ASCollectionView in the following:
```swift
ASCollectionView(...) { ... }
   .forceUpdateWith(self.$layoutInfo)
```

```swift
struct ForceUpdateOnVariableChange<T, Content: View>: View {
	@Binding var variableOfInterest: T
	var content: (() -> Content)
	
	var body: some View {
		content()
	}
}

extension View {
	func forceUpdateWith<T>(_ binding: Binding<T>) -> some View {
		ForceUpdateOnVariableChange(variableOfInterest: binding) {
			self
		}
	}
}
```

**Comment by nferruzzi** [comment]
> ```swift
> .forceUpdateWith(self.$layoutInfo.numberOfColumns)
> ```

tried ! and as strange as it sounds, unfortunately, it makes no difference. But thanks for helping out :)
 
I'm quite sure the problem with my project is not correlated to the ASCollectionView so I think we may close this bug when the patch you made reach master. But at the moment I'm not so sure I needed it in the first place XD

When I see the bug/problem on my project, as you pointed out, it looks like the closure is capturing the variable by value and not by reference (which is strange).
The very same code works fine commenting out an `@EnvironmentObject` from a parent view. What can I say ... sound as a compiler problem because I really don't see any link between the two.

Fortunately for me, I can keep that line commented out and enjoy the layout refreshing properly.
^_^

**Comment by apptekstudios** [comment]
Merged these changes into the master branch. Closing the original issue for now :)

@nferruzzi That is very strange that an environment object elsewhere is causing issues, I think you'd be right about it being some sort of swiftUI or compiler issue.

@tarasis If you're still having the problems regarding layout update as discussed above please open another issue so we can keep working on that 👍

**Comment by tarasis** [comment]
@apptekstudios will do. Haven't had a chance yet to try and distill the project to be able to send you a copy. Certainly still unable to change columns as the data is being loaded.

**Comment by nferruzzi** [comment]
@apptekstudios I found the problem at the end it was me misinterpreting where `ObservedObject` are stored.. they are not like @State that persistent values, they are just wrapped properties that create binds on the fly for keypaths. So if you create them locally once your View is recreated they are recreated too.

Basically all I had to do in my project was moving `layoutState` outside of the View definition, something like this to make it work:

```
var globalLayoutState = LayoutState()

struct AdjustableGridScreen: View
{
	@ObservedObject var layoutState = globalLayoutState
```

now I understand why the parent @EnvironmentObject was the source of problems, it was just doing its job by recreating the hierarchy. Everything now makes perfect sense :D

The whole model management in swiftui based around immutable views should have sounded a bell but it didn't, anyway lesson learned. Thanks for your time.
